First implementation of CUDA cascading libraries in celeritas

pcanal opened the pull request:
This leverage the implementation done for VecGeom against
cmake 3.3 and 3.6.  The patch developped have been applied
to FindCUDA.cmake from cmake 3.19.3 (but since celeritas
does not support that version the code supporting
cmake 3.3 has been removed).

NVidia cascading library ‘rules’:

- call `nvcc -dlink *.cu.o ./needed.a`
- call `g++ —-shared -o …so *.cxx.o *.cu.o dlink.o   ./needed.so`
- call `ar qc .._static.a *.cu.o # NO dlink.o`

DO NOT LINK against `libcudart_static.a` !!!

Note: It might be possible to replace this by a new 'add_library'
wrapper creating an object library and creating the 2 libraries
(static and shared) from those objects.

These changes were originally developed in the context of VecGeom
with the following commits:
```
commit b8dc86c4d2b429c9cf4bad65797f2251a2312dfb
Author: Benjamin Morgan <ben.morgan@warwick.ac.uk>
Date:   Tue Mar 10 07:41:23 2020 +0000

    Use GNUInstallDirs to define install locations

    Replace custom install location variables with standard and FHS
    compilant set provided by CMake's builtin GNUInstallDirs module.

    Provide a basic set of CMake settings to check for absolute install
    locations and to reduce verbosity of install messages when files
    have not changed.

    Replace all occurences (except Jenkins and RecordCmdLine.cmake) of
    CMAKE_{SOURCE,BINARY}_DIR variables with PROJECT_{SOURCE,BINARY}_DIR.
    The PROJECT versions are more robust under the use case that VecGeom
    is imported/built as a submodule of a downstream project. Jenkins
    scripts are left as-is as they are project specific. RecordCmdLine.cmake
    is unchanged as it must run before project() is called and thus the
    PROJECT variables do not exist.

    RecordCmdLine is called before Project

commit 5f3640044654b41aef331a60457936659e769dc7
Author: Philippe Canal <pcanal@fnal.gov>
Date:   Fri Jun 23 15:03:01 2017 -0500

    Cache FindCUDA.cmake variable to avoid re-initialization.

    See https://gitlab.kitware.com/cmake/cmake/issues/17013

commit 740aa1454d1e1debf81cf8ed4aca4fe457db2a8c
Author: Philippe Canal <pcanal@fnal.gov>
Date:   Thu Jun 30 16:39:44 2016 -0500

    GEANT-133 Work around deficiency in CMake handling/substituion of target properties.

    See http://public.kitware.com/Bug/view.php?id=14353 for more detail and most of the work around (we added handling of quotes in the
    arguments).

commit 830d7eb641c4e777c40ed9a776f3544b97e91d8b
Author: Philippe Canal <pcanal@fnal.gov>
Date:   Wed May 11 14:21:38 2016 -0500

    VECGEOM-35 Allow cascading cuda shared libraries

    Add support for cascading shared library in the same way as for static
    library.

    Instead of libname.a, a static library named libname_static.a is created.
    Instead of libname_final.a a shared library named libname.so is created.

    To use, add something like:

      cuda_add_library(cudauserlib ${SRC_USER_CUDA_CPP} SHARED OPTIONS ${CUDA_ARCH} )
      cuda_add_library_depend(cudauserlib vecgeomcuda_static libvecgeomcuda_static.a)

      target_link_libraries(cudauserlib ${VECGEOM_LIBRARIES} ${VECGEOM_LIBRARIES_EXTERNAL})
      set(USERKERNELLIB cudauserlib)

    Note: the repeat in cuda_add_library_depend below is due the author of cuda_add_library_depend
    not knowing how to automatically go from the target to the real file in add_custom_command.

commit 1d7a7e5d32f2938b7de82815f86b03d1c56418e2
Author: Philippe Canal <pcanal@fnal.gov>
Date:   Wed Mar 30 15:25:28 2016 -0500

    VECGEOM-25 Fix dependency ... by forcing the user to repeat ...

commit 012ad8f1e7574893207816576807f98de70b9c81
Author: Philippe Canal <pcanal@fnal.gov>
Date:   Wed Mar 30 13:38:21 2016 -0500

    VECGEOM-35 Allow cascading cuda libraries

    In order to allow cascading cuda libraries, we need to have a simple
    version of the libraries which contains solely the .o files corresponding
    to their content.

    Then once the full set of libraries is known, we need to generate a single
    intermediary 'link' .o file, generated with nvcc with the -dlink options
    that is linked against the .o of the most derived libraries and the 'simple
    version' of the dependent libraries.

    The final useable library is named lib[targetname]_final.a.  To use
    do something like:

      cuda_add_library(cudauserlib ${SRC_USER_CUDA_CPP} OPTIONS ${CUDA_ARCH} )
      cuda_add_library_depend(cudauserlib libvecgeomcuda.a)
      target_link_libraries(cudauserlib ${VECGEOM_LIBRARIES} ${VECGEOM_LIBRARIES_EXTERNAL})
      set(USERKERNELLIB cudauserlib_final)

commit 17ca13ed4b8f350d6e6517205f4e250b1cad520a
Author: Philippe Canal <pcanal@fnal.gov>
Date:   Tue May 6 14:45:02 2014 -0500

    Fix shared library build with individual .o for cuda source file

commit 53830ecd166caf839f49bd72c3aff4d3614d0378
Author: Philippe Canal <pcanal@fnal.gov>
Date:   Tue May 2 13:08:31 2017 -0500

    Import FindCUDA from cmake 3.6.20160902-g1eb3e
```

sethrj commented:
@pcanal Some critical issues:
- The [no-cuda build is failing](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Celeritas/detail/PR-233/1/pipeline/10)
- The [cuda build is failing](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Celeritas/detail/PR-233/1/pipeline/)
- There are several accidentally committed files and a duplicate findcuda cmake
- Is there not a way to get linking working with first-class CUDA support?
- Is it really necessary to force shared libraries for everything?

If this is what has to be done to work with VecGeom in the short term... so be it. But do you agree that this is unsustainable?

pcanal commented:
> The cuda build is failing

Was missing an essential file.  Added.

> The no-cuda build is failing

Straight-forward to fix (either adding and if or another layer of indirection). Fixed.

> There are several accidentally committed files 

Sorry.  Fixed.

> and a duplicate findcuda cmake

That is completely intentional.  If we were not using squash-on-merge (and this was not a temporary solution), I would have made a commit with the "pristine" copy from the cmake distribution as its own commit and the change as a separate commit so that the file could later be updated from a new version of cmake and the patch reapplied).  Having the copy in-repository avoid any ambiguity on what was the exact content of the original file.

> Is there not a way to get linking working with first-class CUDA support?

I think there is "Note: It might be possible to replace this by a new 'add_library'
wrapper creating an object library and creating the 2 libraries
(static and shared) from those objects." but this will take longer to develop.  Since the path taken here is "known" to work, I  took it as the shortest path to a working state.

> Is it really necessary to force shared libraries for everything?

It is not necessary.  I just skipped as a non-essential.  Technically the code I inherited did string manipulation expecting "SHARED" or "STATIC" in the add_library code while the celeritas cmake currently use a global setting (if I understood correctly).  This is fine and could be handled (adding new if statements to deal with the global setting rather the underlying settings), however I would rather invest in a cleaner solution that add feature to the temporary code.

> If this is what has to be done to work with VecGeom in the short term... so be it. But do you agree that this is unsustainable?

I agree that the premise (resurrecting the deprecated/obsolete FindCUDA.cmake) is not sustainable and might not be enough to convince cmake/nvidia to fix the problem upstream.

I am planning on trying out the alternative I sketched in the NOTE but there are some challenges I am sure I will know how to overcome (See the note and commit message part about expand_command), so it not known/knowable when the replacement will be ready.

So for now let's go with this.

pcanal commented:
@sethrj One of the build still fails. I can not see the actual failing command line (eventhough ninja is run in verbose mode, it turns around and call the cmake wrapper in non verbose mode:
```
/opt/software/linux-ubuntu20.04-x86_64/gcc-9.3.0/cmake-3.18.4-liggd6utlx54xhzls3uypu3adngo6xe5/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=Debug -D generated_file:STRING=/var/jenkins/workspace/Celeritas_PR-233/build/src/CMakeFiles/celeritas.dir/sim/detail/./celeritas_generated_SimStateInit.cu.o -D generated_cubin_file:STRING=/var/jenkins/workspace/Celeritas_PR-233/build/src/CMakeFiles/celeritas.dir/sim/detail/./celeritas_generated_SimStateInit.cu.o.cubin.txt -P /var/jenkins/workspace/Celeritas_PR-233/build/src/CMakeFiles/celeritas.dir/sim/detail/celeritas_generated_SimStateInit.cu.o.Debug.cmake
```

sethrj commented:
> That is completely intentional. If we were not using squash-on-merge (and this was not a temporary solution), I would have made a commit with the "pristine" copy from the cmake distribution as its own commit and the change as a separate commit so that the file could later be updated from a new version of cmake and the patch reapplied). Having the copy in-repository avoid any ambiguity on what was the exact content of the original file.
Fortunately, all of the original PR histories are retained on github, both through the web interface and through git (e.g. `git fetch upstream refs/pull/233/head mr-233`), even after merging. That allows us all to pull the "diff" of the FindCUDA.  I'd prefer that (with the annotation of the original CMake version in the git commit history) to an unused duplicate of the CMake file.

> It is not necessary. I just skipped as a non-essential. Technically the code I inherited did string manipulation expecting "SHARED" or "STATIC" in the add_library code while the celeritas cmake currently use a global setting (if I understood correctly). This is fine and could be handled (adding new if statements to deal with the global setting rather the underlying settings), however I would rather invest in a cleaner solution that add feature to the temporary code.
Fair enough. However, since static libraries are nominally supported through the cmake options, I think it would be good to add to the CMake code an assertion:
```cmake
if(NOT BUILD_SHARED_LIBS)
  # TODO: delete this when we support first-class CUDA again
  message(SEND_ERROR "Static build is currently unsupported")
endif()
```

> So for now let's go with this.
Fair enough! 🙂

sethrj commented:
> One of the build still fails. I can not see the actual failing command line

@pcanal I tried reproducing with the default build on emmet but got a billion error messages, which may be what's causing the failure on the CI:
```
[2/345] Building NVCC (Device) object src/C...etail/celeritas_generated_SimStateInit.cu.o
FAILED: src/CMakeFiles/celeritas.dir/sim/detail/celeritas_generated_SimStateInit.cu.o
cd /home/s3j/.local/src/celeritas/build/src/CMakeFiles/celeritas.dir/sim/detail && /projects/spack/opt/spack/gcc-8.3.1/cmake/547luje/bin/cmake -E make_directory /home/s3j/.local/src/celeritas/build/src/CMakeFiles/celeritas.dir/sim/detail/. && /projects/spack/opt/spack/gcc-8.3.1/cmake/547luje/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=Debug -D generated_file:STRING=/home/s3j/.local/src/celeritas/build/src/CMakeFiles/celeritas.dir/sim/detail/./celeritas_generated_SimStateInit.cu.o -D generated_cubin_file:STRING=/home/s3j/.local/src/celeritas/build/src/CMakeFiles/celeritas.dir/sim/detail/./celeritas_generated_SimStateInit.cu.o.cubin.txt -P /home/s3j/.local/src/celeritas/build/src/CMakeFiles/celeritas.dir/sim/detail/celeritas_generated_SimStateInit.cu.o.Debug.cmake
/tmp/tmpxft_003ef740_00000000-5_SimStateInit.cudafe1.cpp:1:3: error: style of line directive is a GCC extension [-Werror]
 # 1 "/home/s3j/.local/src/celeritas/src/sim/detail/SimStateInit.cu"
   ^
```

You can temporarily set the `CUDA_VERBOSE_BUILD=ON` cmake option to show the compiler command.

sethrj commented on app/CMakeLists.txt:
```diff
@@ -143,19 +143,26 @@ endif()
+  # Note: the repeat (target name, location) below is due the author of  cuda_add_library_depend
+  # not knowing how to automatically go from the target to the real file from a generator expression in add_custom_command
+  get_property(vecgeom_static_target_location TARGET VecGeom::vecgeomcuda_static PROPERTY LOCATION)
+  cuda_add_library_depend(celeritas_demo_rasterizer celeritas_static ../src/libceleritas_static.a VecGeom::vecgeomcuda_static ${vecgeom_static_target_location})
```
Identical code shows up in a few places; can this not be rolled into `celeritas_link_vecgeom_cuda`? If not, can you make a separate helper function/macro?

sethrj commented on app/CMakeLists.txt:
```diff
@@ -196,23 +203,33 @@ if(CELERITAS_BUILD_DEMOS AND CELERITAS_USE_VecGeom)
     demo-loop/LDemoParams.cc
     demo-loop/LDemoRun.cc
     ${_cuda_src}
+    SHARED
```
Perhaps put the `SHARED` into the `celeritas_add_library` since it's (for now) temporary?

pcanal replied:
Actually now that we have `celeritas_add_library` it is simpler to add.

sethrj commented on cmake/CeleritasUtils.cmake:
```diff
@@ -42,16 +42,50 @@ macro(celeritas_find_package_config _package)
+
+  GET_TARGET_PROPERTY(target_LINK_LIBRARIES ${target} LINK_LIBRARIES )
+  GET_TARGET_PROPERTY(target_INTERFACE_LINK_LIBRARIES ${target} INTERFACE_LINK_LIBRARIES )
+  SET_TARGET_PROPERTIES(${target}_static PROPERTIES LINK_LIBRARIES "${target_LINK_LIBRARIES}" INTERFACE_LINK_LIBRARIES "${target_INTERFACE_LINK_LIBRARIES}")
```
Even though this code is (hopefully) temporary, could you update the formatting to match the rest of the cmake?
- Lowercase function calls, 
- 80-character lines,
- separate key/value pairs for `set_target_properties` for readability
- Remove dead code rather than commenting out

sethrj commented on cmake/CeleritasUtils.cmake:
```diff
@@ -42,16 +42,50 @@ macro(celeritas_find_package_config _package)
+  GET_TARGET_PROPERTY(target_INTERFACE_LINK_LIBRARIES ${target} INTERFACE_LINK_LIBRARIES )
+  SET_TARGET_PROPERTIES(${target}_static PROPERTIES LINK_LIBRARIES "${target_LINK_LIBRARIES}" INTERFACE_LINK_LIBRARIES "${target_INTERFACE_LINK_LIBRARIES}")
+
+  target_compile_options(${target}_static PRIVATE "-fPIC")
```
```suggestion
  set_target_properties(${target}_static PROPERTIES
    POSITION_INDEPENDENT_CODE ON
  )
```

sethrj commented:
@pcanal This is a code error (missing `inline` keyword) that points to a hole in our testing. I don't think we should worry *too* much about covering all possible configure options, especially given the extra restrictions the VecGeom link is imposing. I'm going to change the `ndebug` CI configuration to use static libraries, which is a good choice anyway for an optimized build and/or when running on a parallel file system. The static build should catch the missing `inline` and hopefully any others lurking.

pcanal commented:
Replaced by https://github.com/celeritas-project/celeritas/pull/239
